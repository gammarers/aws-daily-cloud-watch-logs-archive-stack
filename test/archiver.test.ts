import { App } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { DailyCloudWatchLogsArchiveStack } from '../src';

describe('DailyCloudWatchLogsArchiveStack Testing', () => {
  const app = new App();

  const stack = new DailyCloudWatchLogsArchiveStack(app, 'DailyCloudWatchLogsArchiveStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
    targetResourceTag: {
      key: 'DailyLogExport',
      values: ['Yes'],
    },
  });

  const template = Template.fromStack(stack);

  describe('Bucket Testing', () => {

    it('Should have bucket encryption', () => {
      template.hasResourceProperties('AWS::S3::Bucket', {
        BucketEncryption: Match.objectEquals({
          ServerSideEncryptionConfiguration: [
            {
              ServerSideEncryptionByDefault: {
                SSEAlgorithm: 'AES256',
              },
            },
          ],
        }),
      });
    });

    it('Should have bucket resource policy', () => {
      template.hasResourceProperties('AWS::S3::BucketPolicy', {
        PolicyDocument: {
          Version: '2012-10-17',
          Statement: Match.arrayWith([
            Match.objectEquals({
              Effect: 'Allow',
              Action: 's3:GetBucketAcl',
              Principal: {
                Service: 'logs.us-east-1.amazonaws.com',
              },
              Resource: {
                'Fn::GetAtt': [
                  Match.stringLikeRegexp('LogArchiveBucket.*'),
                  'Arn',
                ],
              },
            }),
            Match.objectEquals({
              Effect: 'Allow',
              Action: 's3:PutObject',
              Principal: {
                Service: 'logs.us-east-1.amazonaws.com',
              },
              Resource: {
                'Fn::Join': [
                  '',
                  [
                    {
                      'Fn::GetAtt': [
                        Match.stringLikeRegexp('LogArchiveBucket.*'),
                        'Arn',
                      ],
                    },
                    '/*',
                  ],
                ],
              },
              Condition: {
                StringEquals: {
                  's3:x-amz-acl': 'bucket-owner-full-control',
                },
              },
            }),
          ]),
        },
      });
    });

  });

  describe('StepFunctions Testing', () => {
    it('Should have Stepfunctions(StateMachine) execution role', () => {
      template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
        RoleName: Match.stringLikeRegexp('daily-cw-logs-archive-machine-.*-role'),
        Description: 'daily CloudWatch Logs archive machine role.',
        AssumeRolePolicyDocument: Match.objectEquals({
          Version: '2012-10-17',
          Statement: Match.arrayEquals([
            Match.objectEquals({
              Effect: 'Allow',
              Principal: {
                Service: 'states.amazonaws.com',
              },
              Action: 'sts:AssumeRole',
            }),
          ]),
        }),
      }));
    });
    it('Should have StepFunctions(StateMachine) default policy', () => {
      template.hasResourceProperties('AWS::IAM::Policy', Match.objectEquals({
        PolicyName: Match.stringLikeRegexp('daily-cw-logs-archive-machine-.*-default-policy'),
        Roles: Match.arrayEquals([
          {
            Ref: Match.stringLikeRegexp('StateMachineRole.*'),
          },
        ]),
        PolicyDocument: Match.objectEquals({
          Statement: Match.arrayEquals([
            {
              Action: 'tag:GetResources',
              Effect: 'Allow',
              Resource: '*',
            },
            {
              Action: 'lambda:InvokeFunction',
              Effect: 'Allow',
              Resource: Match.arrayEquals([
                {
                  'Fn::GetAtt': Match.arrayEquals([
                    Match.stringLikeRegexp('LogArchiveFunction.*'),
                    'Arn',
                  ]),
                },
                {
                  'Fn::Join': Match.arrayEquals([
                    '',
                    [
                      {
                        'Fn::GetAtt': Match.arrayEquals([
                          Match.stringLikeRegexp('LogArchiveFunction.*'),
                          'Arn',
                        ]),
                      },
                      ':*',
                    ],
                  ]),
                },
              ]),
            },
            {
              Action: 'logs:DescribeExportTasks',
              Effect: 'Allow',
              Resource: '*',
            },
          ]),
          Version: '2012-10-17',
        }),
      }));
    });
    it('Should have StateMachine', () => {
      template.hasResourceProperties('AWS::StepFunctions::StateMachine', Match.objectEquals({
        StateMachineName: Match.stringLikeRegexp('daily-cw-logs-archive-.*-machine'),
        DefinitionString: Match.anyValue(),
        RoleArn: Match.objectEquals({
          'Fn::GetAtt': Match.arrayEquals([
            Match.stringLikeRegexp('StateMachineRole.*'),
            'Arn',
          ]),
        }),
      }));
    });
  });

  describe('Lambda Testing', () => {

    it('Should have lambda execution role', () => {
      template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
        RoleName: Match.stringLikeRegexp('daily-cw-logs-archive-lambda-exec-.*-role'),
        Description: Match.anyValue(),
        AssumeRolePolicyDocument: Match.objectEquals({
          Version: '2012-10-17',
          Statement: Match.arrayWith([
            Match.objectEquals({
              Effect: 'Allow',
              Principal: {
                Service: 'lambda.amazonaws.com',
              },
              Action: 'sts:AssumeRole',
            }),
          ]),
        }),
        ManagedPolicyArns: Match.arrayWith([
          {
            'Fn::Join': Match.arrayEquals([
              '',
              Match.arrayEquals([
                'arn:',
                {
                  Ref: 'AWS::Partition',
                },
                ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
              ]),
            ]),
          },
        ]),
        Policies: Match.arrayEquals([
          {
            PolicyName: 'log-export-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: 'logs:CreateExportTask',
                  Resource: '*',
                }),
              ],
            }),
          },
          {
            PolicyName: 'put-bucket-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: Match.arrayEquals([
                    's3:GetBucketAcl',
                    's3:PutObject',
                  ]),
                  Resource: {
                    'Fn::GetAtt': Match.arrayEquals([
                      Match.stringLikeRegexp('LogArchiveBucket.*'),
                      'Arn',
                    ]),
                  },
                }),
              ],
            }),
          },
        ]),
      }));
    });

    it('Should have lambda function', () => {
      template.hasResourceProperties('AWS::Lambda::Function', Match.objectEquals({
        FunctionName: Match.stringLikeRegexp('daily-cw-logs-archive-.*-func'),
        Handler: 'index.handler',
        Runtime: 'nodejs22.x',
        Code: {
          S3Bucket: Match.anyValue(),
          S3Key: Match.stringLikeRegexp('.*.zip'),
        },
        Description: Match.anyValue(),
        Environment: {
          Variables: {
            AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
            BUCKET_NAME: {
              Ref: Match.stringLikeRegexp('LogArchiveBucket.*'),
            },
          },
        },
        Role: {
          'Fn::GetAtt': [
            Match.stringLikeRegexp('LambdaExecutionRole.*'),
            'Arn',
          ],
        },
      }));
    });
  });

  describe('Schedule Testing', () => {
    it('Should have Schedule policy', () => {
      template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
        RoleName: Match.stringLikeRegexp('daily-cw-logs-archive-.*-schedule-role'),
        Description: Match.anyValue(),
        AssumeRolePolicyDocument: Match.objectEquals({
          Version: '2012-10-17',
          Statement: Match.arrayEquals([
            {
              Action: 'sts:AssumeRole',
              Effect: 'Allow',
              Principal: {
                Service: 'scheduler.amazonaws.com',
              },
            },
          ]),
        }),
        Policies: Match.arrayEquals([
          {
            PolicyName: Match.stringLikeRegexp('state-machine-exec-policy'),
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: Match.arrayEquals([
                {
                  Action: 'states:StartExecution',
                  Effect: 'Allow',
                  Resource: {
                    Ref: Match.stringLikeRegexp('StateMachine.*'),
                  },
                },
              ]),
            }),
          },
        ]),
      }));
    });

    it('Should have Schedule', () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('daily-cw-logs-archive-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: Match.stringLikeRegexp('cron(.* 13 * * ? *)'),
        Target: Match.objectEquals({
          Arn: {
            Ref: Match.stringLikeRegexp('StateMachine.*'),
          },
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('SchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: Match.anyValue(),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
      template.resourceCountIs('AWS::Scheduler::Schedule', 1);
    });
  });

  describe('Snapshot Testing', () => {
    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('archiver');
    });
  });
});
