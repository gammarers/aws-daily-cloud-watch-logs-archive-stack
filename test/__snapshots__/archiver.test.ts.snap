// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DailyCloudWatchLogsArchiveStack Testing Snapshot Testing Should match snapshot: archiver 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LambdaExecutionRoleD5C26073": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "daily CloudWatch Logs archive machine exec role.",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "logs:CreateExportTask",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "log-export-policy",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "s3:GetBucketAcl",
                    "s3:PutObject",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "LogArchiveBucketC3E3979A",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "put-bucket-policy",
          },
        ],
        "RoleName": "daily-cw-logs-archive-lambda-exec-fb5caf47-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaFunctionLogGroupD722B4F5": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "LogArchiveFunctionA8152A28",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "LogArchiveBucketC3E3979A": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessControl": "Private",
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "log-archive-fb5caf47",
        "LifecycleConfiguration": {
          "Rules": [
            {
              "Id": "archive-step-lifecycle-rule",
              "Status": "Enabled",
              "Transitions": [
                {
                  "StorageClass": "STANDARD_IA",
                  "TransitionInDays": 400,
                },
                {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 720,
                },
                {
                  "StorageClass": "DEEP_ARCHIVE",
                  "TransitionInDays": 980,
                },
              ],
            },
          ],
        },
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "BucketOwnerEnforced",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "LogArchiveBucketPolicyD4CDCD89": {
      "Properties": {
        "Bucket": {
          "Ref": "LogArchiveBucketC3E3979A",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "LogArchiveBucketC3E3979A",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "LogArchiveBucketC3E3979A",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetBucketAcl",
              "Effect": "Allow",
              "Principal": {
                "Service": "logs.us-east-1.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "LogArchiveBucketC3E3979A",
                  "Arn",
                ],
              },
            },
            {
              "Action": "s3:PutObject",
              "Condition": {
                "StringEquals": {
                  "s3:x-amz-acl": "bucket-owner-full-control",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "logs.us-east-1.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "LogArchiveBucketC3E3979A",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "LogArchiveFunctionA8152A28": {
      "DependsOn": [
        "LambdaExecutionRoleD5C26073",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "HASH.zip",
        },
        "Description": "A function to archive logs s3 bucket from CloudWatch Logs.",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BUCKET_NAME": {
              "Ref": "LogArchiveBucketC3E3979A",
            },
          },
        },
        "FunctionName": "daily-cw-logs-archive-fb5caf47-func",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LambdaExecutionRoleD5C26073",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "Schedule": {
      "Properties": {
        "Description": "daily CloudWatch Logs archive schedule",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "daily-cw-logs-archive-fb5caf47-schedule",
        "ScheduleExpression": "cron(1 13 * * ? *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": {
          "Arn": {
            "Ref": "StateMachine2E01A3A5",
          },
          "Input": "{"tagKey":"DailyLogExport","tagValues":["Yes"]}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "SchedulerExecutionRole0AF092B7",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "SchedulerExecutionRole0AF092B7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "daily CloudWatch Log archive schedule",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "states:StartExecution",
                  "Effect": "Allow",
                  "Resource": {
                    "Ref": "StateMachine2E01A3A5",
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "state-machine-exec-policy",
          },
        ],
        "RoleName": "daily-cw-logs-archive-fb5caf47-schedule-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachine2E01A3A5": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"GetResources","States":{"GetResources":{"Next":"LogGroupExportMap","Type":"Task","ResultPath":"$.Result","ResultSelector":{"TargetResources.$":"$..ResourceTagMappingList[*].ResourceARN"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:resourcegroupstaggingapi:getResources","Parameters":{"ResourceTypeFilters":["logs:log-group"],"TagFilters":[{"Key.$":"$.tagKey","Values.$":"$.tagValues"}]}},"LogGroupExportMap":{"Type":"Map","End":true,"ItemsPath":"$.Result.TargetResources","MaxConcurrency":1,"Iterator":{"StartAt":"GetLogGroupName","States":{"GetLogGroupName":{"Type":"Pass","Parameters":{"TargetLogGroupName.$":"States.ArrayGetItem(States.StringSplit($, ':'), 6)"},"Next":"InvokeLambdaFunction"},"InvokeLambdaFunction":{"Next":"DescribeExportTasks","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "LogArchiveFunctionA8152A28",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"DescribeExportTasks":{"Next":"ExportTaskStatusCheck","Type":"Task","ResultPath":"$.Result","ResultSelector":{"DescribeExportTasksStatus.$":"$.ExportTasks[0].Status.Code"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:cloudwatchlogs:describeExportTasks","Parameters":{"TaskId.$":"$.TaskId"}},"ExportTaskStatusCheck":{"Type":"Choice","Choices":[{"Variable":"$.Result.DescribeExportTasksStatus","StringEquals":"FAILED","Next":"GetLogGroupName"},{"Variable":"$.Result.DescribeExportTasksStatus","StringEquals":"RUNNING","Next":"ExportRunningWait"},{"Variable":"$.Result.DescribeExportTasksStatus","StringEquals":"PENDING","Next":"ExportPendingWait"}],"Default":"Succeed"},"Succeed":{"Type":"Succeed"},"ExportRunningWait":{"Type":"Wait","Seconds":10,"Next":"DescribeExportTasks"},"ExportPendingWait":{"Type":"Wait","Seconds":3,"Next":"DescribeExportTasks"}}}}}}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
        "StateMachineName": "daily-cw-logs-archive-fb5caf47-machine",
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "StateMachineRoleB840431D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "states.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "daily CloudWatch Logs archive machine role.",
        "RoleName": "daily-cw-logs-archive-machine-fb5caf47-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "tag:GetResources",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "LogArchiveFunctionA8152A28",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "LogArchiveFunctionA8152A28",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "logs:DescribeExportTasks",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "daily-cw-logs-archive-machine-fb5caf47-default-policy",
        "Roles": [
          {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
